### PR TITLE
refactor interfaces

### PR DESCRIPTION
**Pull request by magik6k**

### CODE REVIEW

File: basic.go [hunk 1]
```diff
@@ -93,67 +93,83 @@ func (b *basicBus) tryDropNode(typ reflect.Type) {
...
 		if err := opt(&settings); err != nil {
 			return nil, err
 		}
 	}
 
-	refCh := reflect.ValueOf(typedChan)
-	typ := refCh.Type()
-	if typ.Kind() != reflect.Chan {
-		return nil, errors.New("expected a channel")
+	types, ok := evtTypes.([]interface{})
```
Comment (raulk) [inline basic.go:139]: ~if this is a single type (which the interface explicitly allows), `types` will be nil. We need to enclose the single type inside a slice.~

EDIT: Misread.

File: basic.go [hunk 2]
```diff
@@ -93,67 +93,83 @@ func (b *basicBus) tryDropNode(typ reflect.Type) {
...
+	nodes   []*node
+	dropper func(reflect.Type)
+}
+
+func (s *sub) Out() <-chan interface{} {
+	return s.ch
+}
+
+func (s *sub) Close() error {
+	close(s.ch)
```
Comment (raulk) [inline basic.go:107]: ~we should probably make `s.ch` nil to allow for gc to vacuum the channel even if the consumer is retaining a reference to the sub.~

EDIT: we shouldn't do this if we want the channel to be accessible so the consumer can drain it after closing.

File: basic.go [hunk 3]
```diff
@@ -220,11 +236,11 @@ func (n *node) emit(event interface{}) {
 
 	n.lk.RLock()
 	if n.keepLast {
-		n.last.Store(eval)
+		n.last.Store(event)
 	}
 
 	for _, ch := range n.sinks {
-		ch.Send(eval)
+		ch <- event
```
Comment (raulk) [inline basic.go:243]: We'll soon want to handle backpressure here, but it's fine for now.

**Review by raulk — COMMENTED** [review]
~noticed a tiny bug; I'll push a commit to fix it. Other than that,~ LGTM.

EDIT: nevermind, apparently I cannot read negative `if`s.

**Comment by raulk** [comment]
Rectified my review above; apparently I can't read code on a Friday afternoon 😛 I'll release go-libp2p-core, then update go.mod here and merge.

File: basic.go [hunk 4]
```diff
@@ -93,67 +93,83 @@ func (b *basicBus) tryDropNode(typ reflect.Type) {
...
 	}
 
-	if settings.forcedType != nil {
-		if settings.forcedType.Elem().AssignableTo(typ) {
-			return nil, fmt.Errorf("forced type %s cannot be sent to chan %s", settings.forcedType, typ)
+	for i, etyp := range types {
+		typ := reflect.TypeOf(etyp)
+
+		if typ.Kind() != reflect.Ptr {
+			return nil, errors.New("subscribe called with non-pointer type")
```
Comment (Kubuxu) [inline basic.go:155]: This needs to close already created subscriptions.

Comment (raulk) [reply basic.go:155]: 👍 yeah, since this is merged and released, could you open an issue for it?

**Pull request merged**